### DISCUSSION TITLE
templates in nested structs

### DISCUSSION THREAD

**Discussion by georgfaust** [cue-lang/cue#1728] [Q&A] [answered]
templates are very useful to avoid redundancy
```
foo: [F=_]: {f: F}
foo: test: {}
```
but how can I use a template in different levels of a nested struct?
this will not match the template:
```
nested: foo: test: {} 
```
the template has to match the complete path:
```
nested: bar: [B=_]: {b: B}
nested: bar: test: {}
```
Is there a way to write templates that can be used in any level of a nested struct?
```
foobar: test: {}
nested: foobar: test: {}
nested: deep: foobar: test: {}
```

**Comment by myitcv** [comment, marked as answer]
I assume this is a follow-up to https://github.com/cue-lang/cue/discussions/1725?

If so, I'm further assuming you want to constrain the value of the field `f` to be the path from the root level?

If so, this is not currently possible. But I've raised https://github.com/cue-lang/cue/issues/165#issuecomment-1135440557 as a discussion point in the query proposal.

**Reply by georgfaust** [reply]
I do not yet see how this could help in building a path/ID, but it may.

I give you a real world example, why I think this could be generally useful.

Say we have this, `floor` and `part` both can have `rooms`. 
```
floor: [Name=_]: {name: Name}
part: [Name=_]: {name: Name}

floor: basement: room: r1: {}
part: pool: room: r1: {}
```
For the template to match I have to add:
```
floor: [_]: room: [Name=_]: {name: Name}
part: [_]: room: [Name=_]: {name: Name}
```
instead of just
```
room: [Name=_]: {name: Name}
```

**Reply by myitcv** [reply]
Yes, as I've mentioned elsewhere, CUE does not have this concept of "path from the root to the current value"
